### PR TITLE
Add a sample of custom window

### PR DESCRIPTION
**Pull request by jeanlouisboudart**
Here is a basic example of a custom daily window starting every day at 6pm.

It also shows a potential problem with timezone having daylight saving time (https://en.wikipedia.org/wiki/Daylight_saving_time).

### CODE REVIEW

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    private static final String outputTopic = "outputTopic";
+    private static final ZoneId zone = ZoneOffset.UTC;
+
+    @Test
+    public void shouldSumNumbersOnSameDay() {
+        final List<MyEvent> inputValues = Arrays.asList(
+                new MyEvent(1, ZonedDateTime.of(2019, 1, 1, 16, 29, 0, 0, zone)),
+                new MyEvent(2, ZonedDateTime.of(2019, 1, 1, 16, 30, 0, 0, zone)),
+                new MyEvent(7, ZonedDateTime.of(2019, 1, 1, 16, 31, 0, 0, zone)),
+                new MyEvent(7, ZonedDateTime.now())
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: Is this last event to flush everything through suppression? Might be clearer to explicitly provide a record that's at time `window end + grace period` for this purpose.

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                        ZonedDateTime.of(2019, 3, 29, 18, 0, 0, 0, zoneWithDST),
+                        ZonedDateTime.of(2019, 3, 30, 18, 0, 0, 0, zoneWithDST)),
+                new ExpectedResult(1,
+                        10,
+                        ZonedDateTime.of(2019, 3, 30, 18, 0, 0, 0, zoneWithDST),
+                        //This get one extra hour due to time shift in Daylight saving
+                        //As a user i would expect it to end on 31th at 6pm.
+
+                        //The limitation seems to come from TimeWindowSerializer / TimeWindowDeserializer as we serialize only start date.
+                        //Suggestion: By serializing both start and end of window, we could support more complex cases with non fixed time window and address daylight savings on daily windows.
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: Yes, this is exactly right. Feel free to open a ticket for this, and thanks for providing an example of how to accomplish custom time windows that explicitly includes this edge case!

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+import java.time.Duration;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+
+public class CustomWindowTest {
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: These tests are very well structured. Thanks!

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java (hunk 2 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: Also, it's worth noting that days will have different lengths under other conditions than just leap year. Leap seconds, for example.

This is actually part of the reason we use "milliseconds since the epoch" as our clock, and align our time windows to some interval on that clock. That way, the internal clock is both monotonic and regular, neither of which are properties that real-life clocks possess.

Of course, this is an academic choice, since it simplifies the implementation at the expense of realistic use cases like "aggregate daily at midnight".

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                .groupByKey()
+                .windowedBy(new DailyTimeWindows(zoneId, Duration.ofMinutes(30)))
+                // A simple sum of value
+                .reduce((v1, v2) -> v1 + v2, Materialized.with(Serdes.Integer(), Serdes.Integer()))
+                // We only care about final result
+                .suppress(Suppressed.untilWindowCloses(Suppressed.BufferConfig.unbounded()));
+        sumOfOddNumbers.toStream().print(Printed.toSysOut());
+        sumOfOddNumbers.toStream().to(outputTopic, Produced.with(WindowedSerdes.timeWindowedSerdeFrom(Integer.class), Serdes.Integer()));
+
+        TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsConfiguration);
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: Oh, you should be able to use try-with-resources here.

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    private void verifyResults(List<ExpectedResult> expectedValues, TopologyTestDriver testDriver) {
+        for (ExpectedResult expectedValue : expectedValues) {
+            ProducerRecord<Windowed<Integer>, Integer> results = testDriver.readOutput(outputTopic, new TimeWindowedDeserializer(new IntegerDeserializer(), Duration.ofDays(1).toMillis()), new IntegerDeserializer());
+            OutputVerifier.compareKeyValue(results, expectedValue.toWindowed(), expectedValue.value);
+
+        }
+    }
+
+    private void injectFakeData(List<MyEvent> inputValues, TopologyTestDriver testDriver) {
```
Comment (vvcephei) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: There are now some utilities in IntegrationTestUtils that you could use for this and for `verifyResults`.

Comment (jeanlouisboudart) [reply src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: I tried but, it doesn't support providing event with given timestamp yet :)

File: src/test/java/io/confluent/examples/streams/window/DailyTimeWindows.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        this.zoneId = zoneId;
+        this.grace = grace.toMillis();
+    }
+
+    @Override
+    public Map<Long, TimeWindow> windowsFor(long timestamp) {
+        Instant instant = Instant.ofEpochMilli(timestamp);
+
+        ZonedDateTime zonedDateTime = instant.atZone(zoneId);
+        ZonedDateTime startTime = zonedDateTime.getHour() >= 18 ? zonedDateTime.truncatedTo(ChronoUnit.DAYS).withHour(18) : zonedDateTime.truncatedTo(ChronoUnit.DAYS).minusDays(1).withHour(18);
```
[the remark this answers is not in the document]
Comment (jeanlouisboudart) [reply src/test/java/io/confluent/examples/streams/window/DailyTimeWindows.java]: done :)

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+                        //As a user i would expect it to end on 31th at 6pm.
+
+                        //The limitation seems to come from TimeWindowSerializer / TimeWindowDeserializer as we serialize only start date.
+                        //Suggestion: By serializing both start and end of window, we could support more complex cases with non fixed time window and address daylight savings on daily windows.
+                        ZonedDateTime.of(2019, 3, 31, 19, 0, 0, 0, zoneWithDST))
+        );
+        kafkaStreamsMagic(inputValues, expectedValues, zoneWithDST);
+    }
+
+    private void kafkaStreamsMagic(List<MyEvent> inputValues, List<ExpectedResult> expectedValues, ZoneId zoneId) {
```
[the remark this answers is not in the document]
Comment (jeanlouisboudart) [reply src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: done

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java (hunk 1 above)
Comment (jeanlouisboudart) [reply src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: done

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+import org.apache.kafka.streams.test.ConsumerRecordFactory;
+import org.apache.kafka.streams.test.OutputVerifier;
+import org.apache.kafka.test.TestUtils;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.*;
```
Comment (ableegoldman) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: nit: Do we allow wildcard imports here?

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.IntegerDeserializer;
+import org.apache.kafka.common.serialization.IntegerSerializer;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.TopologyTestDriver;
+import org.apache.kafka.streams.kstream.*;
```
Comment (ableegoldman) [inline src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: same

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java (hunk 8 above)
Comment (jeanlouisboudart) [reply src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: I removed star import

File: src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java (hunk 9 above)
Comment (jeanlouisboudart) [reply src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java]: I removed star import

**Comment by vvcephei** [comment]
Looks like the build failed with a linter check that all variables are declared `final`. @jeanlouisboudart , if you want to get a quick fix for this (and if you're using IDEA or another IDE), you can enable a code validation in your IDE to warn you when variables could be final. Then, it'll prompt you for all these locations, and also let you click a button to fix them.

**Comment by vvcephei** [comment]
I've merged to `5.2.1-post`, and done the downstream merge to the later branches. `master` is currently broken until a 5.4.0-SNAPSHOT is published for our dependencies, but I'm running a final verification on `5.3.x` to make sure the downstream merge didn't break anything.

Thanks for the contribution, @jeanlouisboudart !

File: src/test/java/io/confluent/examples/streams/window/DailyTimeWindows.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    public Map<Long, TimeWindow> windowsFor(final long timestamp) {
+        final Instant instant = Instant.ofEpochMilli(timestamp);
+
+        final ZonedDateTime zonedDateTime = instant.atZone(zoneId);
+        final ZonedDateTime startTime = zonedDateTime.getHour() >= startHour ? zonedDateTime.truncatedTo(ChronoUnit.DAYS).withHour(startHour) : zonedDateTime.truncatedTo(ChronoUnit.DAYS).minusDays(1).withHour(startHour);
+        final ZonedDateTime endTime = startTime.plusDays(1);
+
+        final Map<Long, TimeWindow> windows = new LinkedHashMap<>();
+        windows.put(toEpochMilli(startTime), new TimeWindow(toEpochMilli(startTime), toEpochMilli(endTime)));
+        return windows;
```
Comment (peterjurkovic) [inline src/test/java/io/confluent/examples/streams/window/DailyTimeWindows.java]: Does it make sense to cache a result of this method instead of creating new LinkedHashMap + TimeWindow on every invocation?

Comment (mjsax) [reply src/test/java/io/confluent/examples/streams/window/DailyTimeWindows.java]: That might be possible. Not sure how much you gain though. I would only consider it, if you know it's a performance issue. The question is also, how many windows do you cache?  While time progressed, the set of "hot" windows changes.
